packages/authorization/src/decorators/authorize.ts

```diff
@@ -0,0 +1,169 @@
+  AuthorizationMetadata
+> {
+  protected mergeWithOwn(
+    ownMetadata: MetadataMap<AuthorizationMetadata>,
+    target: Object,
+    methodName?: string,
+    // tslint:disable-next-line:no-any
+    methodDescriptor?: TypedPropertyDescriptor<any> | number,
+  ) {
+    ownMetadata = ownMetadata || {};
+    let methodMeta = ownMetadata[methodName!];
+    if (!methodMeta) {
+      methodMeta = {};
+      ownMetadata[methodName!] = methodMeta;
+    }
+    if (this.spec.allowedRoles) {
+      methodMeta.allowedRoles = this.merge(
+        methodMeta.allowedRoles,
+        this.spec.allowedRoles,
+      );
+    }
+    if (this.spec.deniedRoles) {
+      methodMeta.deniedRoles = this.merge(
+        methodMeta.deniedRoles,
+        this.spec.deniedRoles,
+      );
+    }
+    if (this.spec.scopes) {
+      methodMeta.scopes = this.merge(methodMeta.scopes, this.spec.scopes);
+    }
+    if (this.spec.voters) {
+      methodMeta.voters = this.merge(methodMeta.voters, this.spec.voters);
+    }
+
+    if (this.spec.resource) {
+      methodMeta.resource = this.spec.resource;
+    }
+
+    return ownMetadata;
+  }
+
+  private merge<T>(src?: T[], target?: T[]): T[] {
+    const list: T[] = [];
+    const set = new Set<T>(src || []);
+    if (target) {
+      for (const i of target) {
+        set.add(i);
+      }
+    }
+    for (const i of set.values()) list.push(i);
+    return list;
+  }
+}
+/**
+ * Decorator `@authorize` to mark methods that require authorization
+ *
+ * @param spec Authorization metadata
+ */
+export function authorize(spec: AuthorizationMetadata) {
+  return AuthorizeMethodDecoratorFactory.createDecorator(
+    AuthorizationBindings.METADATA,
+    spec,
+  );
+}
+
+export namespace authorize {
+  /**
+   * Shortcut to configure allowed roles
+   * @param roles
+   */
+  export const allow = (...roles: string[]) => authorize({allowedRoles: roles});
+  /**
+   * Shortcut to configure denied roles
+   * @param roles
+   */
+  export const deny = (...roles: string[]) => authorize({deniedRoles: roles});
+  /**
+   * Shortcut to specify access scopes
+   * @param scopes
+   */
+  export const scope = (...scopes: string[]) => authorize({scopes});
+
+  /**
+   * Shortcut to configure voters
+   * @param voters
+   */
+  export const vote = (...voters: (Voter | BindingAddress<Voter>)[]) =>
+    authorize({voters});
+
+  /**
+   * Allows all
+   */
+  export const allowAll = () => allow(EVERYONE);
+
+  /**
+   * Allow all but the given roles
+   * @param roles
+   */
+  export const allowAllExcept = (...roles: string[]) =>
+    authorize({
+      deniedRoles: roles,
+      allowedRoles: [EVERYONE],
+    });
+
+  /**
+   * Deny all
+   */
+  export const denyAll = () => deny(EVERYONE);
+
+  /**
+   * Deny all but the given roles
+   * @param roles
+   */
+  export const denyAllExcept = (...roles: string[]) =>
+    authorize({
+      allowedRoles: roles,
+      deniedRoles: [EVERYONE],
```

Comment:
```ts
export const allowAllExcept = (...roles: string[]) =>
    authorize({
      deniedRoles: roles,
      allowedRoles: [EVERYONE],
    });

export const denyAllExcept = (...roles: string[]) =>
    authorize({
      allowedRoles: roles,
      deniedRoles: [EVERYONE],
    });
```

I noticed that `allowAllExcept` is setting `deniedRoles` first and `allowedRoles` second, while `denyAllExcept` sets `allowedRoles` first. Is the order of these two properties significant? If not, can we use the same order in all helpers please?

Reply:
@bajtos Good point. 
I think the sequence matters, it is how authorizer understands the precedence of rules, like 
- for `authorize.allowAllExcept()`, `deniedRoles` has higher priority than `allowedRoles`
- for `authorize.denyAllExcept()`, `allowedRoles` has higher priority than `deniedRoles`

Reply:
@bajtos I also discussed with @raymondfeng a little bit about the rule precedence, since it's complicated and highly depends on the authorizer's implementation, we'd better discuss in a separate issue.